massa-pool-worker/src/denunciation_pool.rs

```diff
@@ -36,18 +56,159 @@ impl DenunciationPool {
     }
     */
 
-    /// Add a list of denunciation to the pool
-    pub(crate) fn add_denunciation(&mut self, denunciation: Denunciation) {
-        if !Denunciation::is_expired(
-            denunciation.get_slot(),
-            &self.last_cs_final_periods,
-            self.config.denunciation_expire_periods,
-        ) {
-            let de_idx = DenunciationIndex::from(&denunciation);
-            self.denunciations_cache.insert(de_idx, denunciation);
+    pub fn add_denunciation_precursor(&mut self, denunciation_precursor: DenunciationPrecursor) {
+        let slot = denunciation_precursor.get_slot();
+
+        // Do some checkups before adding the denunciation precursor
+
+        let now = MassaTime::now().expect("could not get current time");
+
+        // get closest slot according to the current absolute time
+        let slot_now = get_closest_slot_to_timestamp(
+            self.config.thread_count,
+            self.config.t0,
+            self.config.genesis_timestamp,
+            now,
+        );
+
+        // let last_final_periods = self.channels.pool.get_final_cs_periods();
+        if slot.period
+            < self.last_cs_final_periods[slot.thread as usize]
+                .saturating_sub(self.config.denunciation_expire_periods)
+        {
+            // too old - cannot be denounced anymore
+            return;
+        }
+
+        if slot.period.saturating_sub(slot_now.period) > self.config.denunciation_expire_periods {
+            // too much in the future - ignored
+            return;
+        }
+
+        // Get selected address from selector and check
+        // Note: If the public key of the header creator is not checked to match the PoS,
+        //       someone can spam with headers coming from various non-PoS-drawn pubkeys
+        //       and cause a problem
+        let selected_address = self.selector.get_producer(*slot);
```

Comment:
warning: this checks against the block producer. If it's an endorsement denunciation this check will be wrong

Reply:
Fixed.